fontc/src/main.rs

```diff
@@ -143,33 +156,93 @@ fn glyphs_deleted<'a>(current_inputs: &Input, prev_inputs: &'a Input) -> HashSet
         .collect()
 }
 
-fn main() -> Result<(), Error> {
-    env_logger::init();
+fn add_glyph_work(context: &CompileContext, work: &mut Vec<Box<dyn Work<()>>>) {
+    // Destroy IR for deleted glyphs
+
+    work.extend(
+        glyphs_deleted(&context.current_inputs, &context.prev_inputs)
+            .iter()
+            .map(|glyph_name| DeleteWork::create(context.paths.glyph_ir_file(glyph_name))),
+    );
+
+    // Generate IR for changed glyphs
+    // You'd think we could do source => work => do-in-|| in one go but that angers the compiler
+    work.extend(glyphs_changed(context).into_iter().map(|glyph_name| {
+        context.ir_source.create_glyph_ir_work(
+            glyph_name,
+            context.current_inputs.glyphs.get(glyph_name).unwrap(),
+        )
+    }));
+}
+
+fn do_work(work: Vec<Box<dyn Work<()>>>) -> Result<(), Error> {
+    let work_units = work.len();
+    debug!("{} work units to execute", work_units);
+    let errors: Vec<Result<(), WorkError>> = work
+        .into_par_iter()
+        .map(|work| work.exec())
+        .filter(|r| r.is_err())
+        .collect();
+    for error in errors.iter() {
+        warn!("{:#?}", error);
+    }
+    if !errors.is_empty() {
+        error!(
+            "{}/{} work units failed, details logged at warn",
+            errors.len(),
+            work_units
+        );
+        return Err(Error::IrGenerationError);
+    } else {
+        debug!("{}/{} work units successful", work_units, work_units);
+    }
+    Ok(())
+}
 
-    let build_dir = establish_build_dir();
-    let (config, prev_inputs) = init(&build_dir, Args::parse()).map_err(Error::IoError)?;
+struct CompileContext {
+    paths: Paths,
+    ir_source: Box<dyn Source>,
+    prev_inputs: Input,
+    current_inputs: Input,
+}
+
+fn establish_context(paths: Paths, args: Args) -> Result<CompileContext, Error> {
```

Comment:
this feels like it would make sense as a type method/constructor?

Reply:
Hm. I strongly agree! Will fix.